# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/base/Document.cpp
# Commit: fb89eb43b3ce
# Full Hash: fb89eb43b3ce78c5031edb49a2e62ddfa7988415
# Author: Timothy Nikkel <tnikkel@gmail.com>
# Date: 2024-02-23 03:40:30
# Description:
#   Bug 1836801. If the order of top layer items changes make sure to mark the frame modified. r=gfx-reviewers,aosmond
#   
#   The order of items in the mTopLayer array determines the order that we build display items, so if it changes we need to mark the frames as modified so that retained display list partial updates can properly merge.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D201214
# ==============================================================================

diff -r c706f3410c62 -r fb89eb43b3ce dom/base/Document.cpp
--- a/dom/base/Document.cpp	Thu Feb 22 09:44:39 2024 +0000
+++ b/dom/base/Document.cpp	Thu Feb 22 10:09:18 2024 +0000
@@ -14751,6 +14751,10 @@
   const bool modal = aElement.State().HasState(ElementState::MODAL);
 
   TopLayerPop(aElement);
+  if (nsIFrame* f = aElement.GetPrimaryFrame()) {
+    f->MarkNeedsDisplayItemRebuild();
+  }
+
   mTopLayer.AppendElement(do_GetWeakReference(&aElement));
   NS_ASSERTION(GetTopLayerTop() == &aElement, "Should match");
 
@@ -14804,6 +14808,9 @@
     nsCOMPtr<Element> element(do_QueryReferent(mTopLayer[i]));
     if (element && aPredicate(element)) {
       removedElement = element;
+      if (nsIFrame* f = element->GetPrimaryFrame()) {
+        f->MarkNeedsDisplayItemRebuild();
+      }
       mTopLayer.RemoveElementAt(i);
       break;
     }
@@ -14818,6 +14825,12 @@
   while (!mTopLayer.IsEmpty()) {
     Element* element = GetTopLayerTop();
     if (!element || element->GetComposedDoc() != this) {
+      if (element) {
+        if (nsIFrame* f = element->GetPrimaryFrame()) {
+          f->MarkNeedsDisplayItemRebuild();
+        }
+      }
+
       mTopLayer.RemoveLastElement();
     } else {
       // The top element of the stack is now an in-doc element. Return here.